File: stdin_parent.py
from __future__ import print_function

import sys
from subprocess import Popen, PIPE
from time import sleep

print("Parent starting child", file=sys.stderr)
child = Popen(["node", "stdin_child.js"], stdout=PIPE, stdin=PIPE)
print("Parent started the child, reading the port number", file=sys.stderr)
port = int(child.stdout.readline())
print("Parent received port number from child:", port, file=sys.stderr)
print("Parent goes to sleep", file=sys.stderr)
sleep(2)
if len(sys.argv) == 2 and sys.argv[1] == "crash":
    print("Parent crashing", file=sys.stderr)
    raise Exception
else:
    print("Parent closing child's stdin", file=sys.stderr)
    child.stdin.close()
    print("Parent waits for child to exit", file=sys.stderr)
    child.wait()
    print("Parent exiting", file=sys.stderr)

File: stdin_parent.rb
STDERR.puts "Parent starting child"
in_r, in_w = IO.pipe()
out_r, out_w = IO.pipe()
child = spawn("node", "stdin_child.js", :in=>in_r, :out=>out_w)
STDERR.puts "Parent started the child, reading the port number"
port = out_r.readline
STDERR.print "Parent received port number from child: ", port
STDERR.puts "Parent goes to sleep"
sleep 2
if ARGV.length == 1 and ARGV[0] == "crash"
    STDERR.puts "Parent crashing"
    Process.kill("SEGV", Process.pid)
    STDERR.puts "Parent should not print it, should crash"
else
    STDERR.puts "Parent closing child's stdin"
    in_w.close
    STDERR.puts "Parent waits for child to exit"
    Process.wait child
    STDERR.puts "Parent exiting"
end
